FilterInterface component

ZEDz318 opened the pull request.

flicksolutions commented on src/app.postcss:
```diff
@@ -17,3 +17,80 @@ body {
+	padding-top: 2%;
+	flex-direction: column;
+}
+.container h2 {
```
please try to use tailwind-classes to set the padding. do not change the size of the h2 class. (h2 already exists in tailwind)

ZEDz318 replied:
deal, I changed all this file and used tailwind wherever i could

flicksolutions commented on src/app.postcss:
```diff
@@ -17,3 +17,80 @@ body {
+
+.dark-container {
+	width: 100%;
+	background-color: #333333;
```
colors will work differently (with a variable) but I have to figure it out first (not yet sure how to do it with dark-mode)
so for now: do not change any colors, I'll do that in a next step

ZEDz318 replied:
I left the color for the moment

flicksolutions replied:
I created a Component _ContentContainer_ which you should use with the prop _dark_. So you don't need that class any more.

`<ContentContainer dark>`

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,151 @@
+<script>
+// @ts-nocheck
```
do not use Typescript and then just disable the checks at the beginning.
This project uses JS-Doc comments for type checking. Although I will not be very strict and because it's a small project, don't care too much about the types here.
so: at least remove all the typescript. if you want, add JsDoc equivalents vor the type.

ZEDz318 replied:
Honestly, I tried to do so. I know how to add typescript but not how to remove it .
I tried again just now and even when I remove dependencies and all mentions of it in config files, i still get type errors.... Any advice or direction you can point me to?

flicksolutions replied:
See my newest remarks to fix this

flicksolutions replied:
```suggestion

```

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,151 @@
+
+  onMount(async () => {
+    try {
+      const response = await fetch(filePath);
```
not how we will be getting the data...
Normally you'd do a server-route, but I'm not sure if we have to do it with a store... it can stay like that for now.

ZEDz318 replied:
ok, if or when we need to change it tell me

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,151 @@
+                  </TreeViewItem>
+                {/each}
+              {/if}
+              {#if children && children.length > 0}
```
your doing the same {if} twice in a row ;-) get rid of one :-)

ZEDz318 replied:
ohh yea, didn't see that

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,151 @@
+                          </TreeViewItem>
+                        {/each}
+                      {/if}
+                      {#if innerChildren && innerChildren.length > 0}
```
same here

ZEDz318 replied:
yep, i took it out. 
the entire thing is also noisy and only reads 3 levels into the json, so depending on the data we are expecting, tell me and I'll change it

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,143 @@
+  // Function to convert the tree data into a format suitable for rendering
+  /**
+	 * @param {{ [s: string]: any; } | ArrayLike<any>} data
+	 */
```
```suggestion
	 * @returns {any[]}
```

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,143 @@
+  <h2 class ="h2 pl-16" style="text-align: left;" >Virtueller Katalog</h2>
+  <div class="side_by_side w-screen pl-2 pt-5 flex">
+    <div class="tree w-1/2 box-border text-sm">
+      <TreeView bind:selection={selectedItems} multiple>
```
here the TS-Error actually shows a mistake you made!
This is a boolean prop. you cannot put in an array of something...

flicksolutions replied:
```suggestion
      <TreeView selection multiple>
```

flicksolutions reviewed (changes requested):
Things still to do:
- Fix TS-Problems
- The Tree is a [relational Tree](https://www.skeleton.dev/components/tree-views#relational-selection). Please enable this.
Thanks!

flicksolutions commented on src/app.postcss:
```diff
@@ -12,8 +12,15 @@ body {
+	background-color: #333333;
+	color: #F5F8D5;
+}
+.searchView input{
```
this is also targetable by tailwind. no need to add a class here

flicksolutions commented on src/app.postcss:
```diff
@@ -12,8 +12,15 @@ body {
+	color: #F5F8D5;
+}
+.searchView input{
+	background-color: #F5F8D5 !important;
```
don't use color values. only use primary, secondary or tertiary colors like described here:
https://www.skeleton.dev/docs/colors

do not use !Important unless absolutely necessary.

flicksolutions commented on src/routes/+page.svelte:
I completely reworked this file. please pull the newest commit from the main branch and integrate your component neatly.

ZEDz318 replied:
I incased the search and relateditems in a form tag, but a normal div would work.
I just followed the link you sent me for tailwind forms and thus used forms instead

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+  import { nodes } from '../treeContent';
+    
+let myTreeViewNodes: TreeViewNode[] = nodes;
+let checkedNodes : string[] = ['javascript'];
```
I guess _javascript_ is left over from copy-pasting. please tidy up.

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+    
+let myTreeViewNodes: TreeViewNode[] = nodes;
+let checkedNodes : string[] = ['javascript'];
+let indeterminateNodes : string[] = ['programming', 'language'];
```
tidy up

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+<script lang="ts">
```
there is still some typescript... remove it and us JSDoc instead.
```suggestion
<script>
```

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+	import { RecursiveTreeView, type TreeViewNode } from '@skeletonlabs/skeleton';
+  import { nodes } from '../treeContent';
+    
+let myTreeViewNodes: TreeViewNode[] = nodes;
```
```suggestion
/** @typedef {import('@skeletonlabs/skeleton').TreeViewNode} TreeViewNode */

/**
 * @type {TreeViewNode[]}
* @description An array of nodes.
*/
let myTreeViewNodes = nodes;
```

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+
+<div class="dark-container w-screen flex pt-10 flex-col">
+  <h2 class ="h2 pl-16" style="text-align: left;" >Virtueller Katalog</h2>
+  <div class="side_by_side w-screen pl-2 pt-5 flex">
```
```suggestion
  <div class="pl-2 pt-5 flex">
```

you shouldn't use _w-screen_ because we have paddings in parent containers.
_side\_by\_side_ is a zombie-class I think? doesn't do anything. also: we normally use normal dashes, not single underscores, when naming css classes.

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+											
+</script>
+
+<div class="dark-container w-screen flex pt-10 flex-col">
```
you probably can get rid of this whole div when you put the component inside a ContentContainer

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+<div class="dark-container w-screen flex pt-10 flex-col">
+  <h2 class ="h2 pl-16" style="text-align: left;" >Virtueller Katalog</h2>
+  <div class="side_by_side w-screen pl-2 pt-5 flex">
+    <div class="tree w-1/2 box-border text-sm">
```
```suggestion
    <div class="w-1/2 box-border text-sm">
```

_tree_ is a zombie class. please tidy up!

flicksolutions commented on src/treeContent.ts:
this shouldn't be a .ts file! Please, use Javascript. make it a .js file.

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+      <RecursiveTreeView selection multiple relational {nodes} bind:checkedNodes={checkedNodes} bind:indeterminateNodes />
+      
+    </div>
+    <div class="search w-1/2 border-box">
```
delete this whole div and apply w-1/2 to the form.

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+      
+    </div>
+    <div class="search w-1/2 border-box">
+      <form class='form'>
```
```suggestion
      <form>
```
form doesn't apply any styles?

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+    </div>
+    <div class="search w-1/2 border-box">
+      <form class='form'>
+        <label class="searchView mr-10">
```
searchView doesn't apply any styles? also, CSS classes normally don't use CamelCase. They use hyphens

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+    <div class="search w-1/2 border-box">
+      <form class='form'>
+        <label class="searchView mr-10">
+            <input class="input text-black p-6 text-base placeholder-black" type="text" placeholder="Dies ist der Surchtext..." />
```
```suggestion
            <input class="input text-primary-500 p-6 placeholder-primary-500" type="text" placeholder="Dies ist der Suchtext..." />
```

flicksolutions replied:
(this was ignored) Maybe you didn't see this comment?

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+        <label class="searchView mr-10">
+            <input class="input text-black p-6 text-base placeholder-black" type="text" placeholder="Dies ist der Surchtext..." />
+        </label>
+        <span class="flex pl-6 pt-6 text-base">Suchen in</span>
```
```suggestion
        <span class="flex pl-6 pt-6">Suchen in</span>
```
what's the reason for _text-base_ everywhere?

flicksolutions replied:
Also, I think you can use <p> here. or a better semantic tag instead of _span_ that doesn't mean anything.

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+            <input class="input text-black p-6 text-base placeholder-black" type="text" placeholder="Dies ist der Surchtext..." />
+        </label>
+        <span class="flex pl-6 pt-6 text-base">Suchen in</span>
+        <div class='categorySelect flex text-sm pl-5'>  
```
tidy up the unused classes

flicksolutions replied:
in fact: a cleaner solution would be this: 
```suggestion
        <div class="flex flex-wrap gap-x-2 text-sm pl-5">  
```

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+        </label>
+        <span class="flex pl-6 pt-6 text-base">Suchen in</span>
+        <div class='categorySelect flex text-sm pl-5'>  
+            <div class = 'one-forth w-1/2 border-box'>
```
with the proposed solution you can loose the wrapping divs
```suggestion

```

flicksolutions commented on src/routes/FilterInterface.svelte:
```diff
@@ -0,0 +1,71 @@
+        <span class="flex pl-6 pt-6 text-base">Suchen in</span>
+        <div class='categorySelect flex text-sm pl-5'>  
+            <div class = 'one-forth w-1/2 border-box'>
+                <label class='flex items-center space-x-2'>
```
```suggestion
                <label class="flex items-center space-x-2 w-1/4">
```
looks better, is responsive and uses less code.

additionally, please use an `{#each}` block and create the labels and input like so. this will further increase the readability and shorten the code.

flicksolutions reviewed (changes requested):
You can see, that there still is a lot to do here until we can merge it into main.
When you finished working, please make sure there are no conflicts so we could actually merge it the next time.

ZEDz318 committed:
Added the Filegrid component

I managed to use JDocs insted of ts
tell me if this is not how you want treeContent to be
also, I think I think I should have added it to the hierarchic.svelte? if so, tell me and i will replace

ZEDz318 committed:
Minor change

for some reason, it doesn't render when i use myTreeViewNodes...

ZEDz318 committed:
loaded tree as a prop from pages instead of in the component itself

flicksolutions commented:
I suspect this was closed by accident. or are you creating a new PR and want to flush this one @ZEDz318 ?
I'll reopen it. When you are finished with everything, you can ask for my review again. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review

ZEDz318 committed:
Virtueller Katalog component update

added the search input
added the checkbox filters

ZEDz318 commented on src/lib/components/filegrid.svelte:
I added this type to the JDocs but i am not sure of it
it was working well without it, but thought that this is the more "correct" way to do it

ZEDz318 commented on src/lib/components/search.svelte:
I don't know what you plan for this so I added it alone in a file, but I can merge it with RelatedItems.svelte

flicksolutions replied:
It definetly doesn't belong into RelatedItems.svelte. But it also doesn't need it's own component (for now). So please just put it in +page.svelte

flicksolutions commented on src/lib/components/search.svelte:
```diff
@@ -0,0 +1,3 @@
+ <label>
+    <input class="input text-black p-6 placeholder-black" type="text" placeholder="Dies ist der Surchtext..." />
```
I think it's the third time that I'm suggesting this:
```suggestion
    <input class="input text-black p-6 placeholder-black" type="text" placeholder="Dies ist der Suchtext..." />
```

flicksolutions replied:
also I please don't use black as colors but primary-500 (as I suggested 2 weeks ago):
```suggestion
    <input class="input text-primary-500 p-6 placeholder-primary-500" type="text" placeholder="Dies ist der Suchtext..." />
```

ZEDz318 replied:
Sorry about that...

flicksolutions commented on src/lib/components/RelatedItems.svelte:
This is absolutely not how svelte works.
In short: There is no need to create a fancy generator function. Svelte does stuff like this for you. It's incredibly rare in a svelte-project that you have to construct a html-like-string and then parse it as html.
Please use the concept of the {#each} block: https://learn.svelte.dev/tutorial/each-blocks
When you use each-blocks you'll see, that you might not even need to outsource this into it's own component! do it directly in the +page.svelte please.

Also you missed a subtitle in the design that was there before. 
![grafik](https://github.com/DHBern/becher-frontend/assets/33655122/072561fd-899f-445c-8abc-bb12e1ed79de)

please add that aswell

ZEDz318 replied:
Idk why i didn't think of that.
I changed it and adapted it to mobile as well. 
I had it as flex-col on mobile and flex-row on sm+ since it looked pretty crammed. tell me if you want me to change it.

also, on landscape, the mobile version acts like a desktop so tell me if you want me to change that as well.

flicksolutions commented on src/lib/components/filegrid.svelte:
```diff
@@ -0,0 +1,30 @@
+<RecursiveTreeView selection multiple relational {nodes} bind:checkedNodes={checkedNodes} bind:indeterminateNodes={indeterminateNodes} />
+</div>
+<!-- To show checkedNodes & indeterminateNodes -->
+<!-- <RecursiveTreeView selection multiple relational {nodes} bind:checkedNodes={checkedNodes} bind:indeterminateNodes={indeterminateNodes} on:click={handleClick} /> -->
```
I think you can remove the comments. it should be obvious enough, that checkedNodes and indeterminateNodes are accessible.

ZEDz318 replied:
Just had it there for testing

flicksolutions commented on src/lib/components/filegrid.svelte:
I think you can integrate this into +page.svelte aswell as it's only about 10 lines of code...
We might create a new component for the whole filtering in the future. but for now you can integrate it in +page.svelte

ZEDz318 replied:
done

flicksolutions commented on src/lib/treeContent.js:
id's should be unique. Let's use integers as datatype for ids.

flicksolutions commented on src/routes/+page.svelte:
without giving you much tipps here:
make it look nice on a mobile resolution.
![grafik](https://github.com/DHBern/becher-frontend/assets/33655122/4a57a997-c83f-4527-8447-17653dca3de7)

and remember to style it mobile-first. This means you apply the styles for a mobile view first and the use the md:, lg:, and other breakpoint modifiers.
But maybe you don't even need many breakpoint modifiers and you can do some magic with flex or grid?

ZEDz318 replied:
I looked into this and yes, it would be better if i style it for mobile first.
I'll see what i can do regarding the breakpoints

flicksolutions reviewed (changes requested):
still quite a long way to go I'm afraid.

ZEDz318 committed:
Fixes and Changes to Virtueller Katalog div

Removed Filegrid and search components and added them directly to +pages
Changed relatedItems component
Id fix in treeContetnt

ZEDz318 reviewed (commented):
I added arbitrary numerical values for the json. tell me if you want them in another way.
I also added mobile view, let me know what you think .If you want me to change the first ContentContainer as well let me know.

flicksolutions commented on src/routes/+page.svelte:
```diff
@@ -22,7 +40,21 @@
-	<h2 class="h2">Virtueller Katalog</h2>
+	<h2 class="h2 mb-4">Virtueller Katalog</h2>
+
+	<div class="flex flex-col sm:flex-row">
```
I would change layout at _md_, _sm_ seems to narrow to me. but it's fine

flicksolutions commented on src/routes/+page.svelte:
```diff
@@ -22,7 +40,21 @@
+	<h2 class="h2 mb-4">Virtueller Katalog</h2>
+
+	<div class="flex flex-col sm:flex-row">
+		<div class = 'w-full sm:w-1/2 mb-4 sm:mb-0'>
```
I would only set _max-width_ and not the width itself.

flicksolutions reviewed (approved):
This is awesome!
I can see some problems that might arrive in the future with how deep the actual category tree will be (probably like 5 levels - pretty crazy) and then we might not have enough width to account for all the levels - but we'll fix it as the problem will arise ;-)
For now this is fine and we'll use it! Thank you!
